openmmtools/alchemy.py

```diff
@@ -1984,91 +2323,155 @@ def check_energy_expression(custom_force, parameter):
                         found = True
                         break
             return found
-
+        
         force_labels = {}
-        nonbonded_forces = []
-        sterics_bond_forces = []
-        electro_bond_forces = []
+        nonbonded_forces = {'' :[], 'zero': [], 'one': [], 'two': []}
+        sterics_bond_forces = {'' :[], 'zero': [], 'one': [], 'two': []}
+        electro_bond_forces = {'' :[], 'zero': [], 'one': [], 'two': []}
```

Comment:
It may be possible to avoid hardcoding these by having `check_parameters` finding the region name with something like
```python
if parameter_name.startswith(parameter):
    region_name = parameter_name[len(parameter)+1:] if len(parameter_name) > len(parameter) else ''
    return True, region_name
```

Reply:
Do you thin this can remove all the hard coding in '_find_force_components' or just the lines you highlighted?

Reply:
I think it could remove the hard coding in `_find_force_components` altogether. You'd have to substitute this pattern
```python
if check_energy_expression(force, 'lambda_sterics'):
    if check_energy_expression(force, 'lambda_sterics_zero'):
        sterics_bond_forces['zero'].append([force_index, force])
    ...
```
with something like
```python
if check_energy_expression(force, 'lambda_sterics'):
    # region_name is None if the parameter is not found in the force and '' if it has no suffix.
    region_name = check_parameter(force, 'lambda_sterics_zero')
    sterics_bond_forces[region_name].append([force_index, force])
```
unless there's somehting I don't see.

Reply:
Ah! Sorry, I remember now that I used `check_energy_expression` for the alchemical-alchemical custom forces that are decoupled instead of annihilated (and have no `lambda_sterics` parameter). To make it work we'll have to modify `check_energy_expression` too, not just `check_parameter`. I'm thinking using regex.

Reply:
Something like this works for me (see https://stackoverflow.com/questions/15340582/python-extract-pattern-matches)
```python
>>> a = 'asdfa sdfalambda_sterics_zero*asdf'
>>> import re
>>> p = re.compile(r'lambda_sterics_([A-Za-z0-9]+)')
>>> p.search(a).group(0)
'lambda_sterics_zero'
>>> p.search(a).group(1)
'zero'
```
I think you can do a simple `parameter in custom_force.getEnergyFunction()` as it is done now, and then try that regex to see if it has a region associated.

Reply:
I've reworked `_find_force_components` to remove the hard coded region names. It's not final and is causing HostGuestExplicit with exact PME to fail currently. But is this rework similar to what you were thinking?

Reply:
Finished off `_find_force_components` think it's working as intended. Let me know if there are any more changes that need to be made.